Add contribution_to_project.rst (#20)
ansible/community-docs: contribution_to_project.rst

***********************
Contributing to Ansible
***********************

This document describes many different ways you can contribute to the Open Source projects of the Ansible ecosystem.

.. contents::
   :local:

Finding the right contribution type for you
-------------------------------------------

A contribution to an Open Source project is any positive addition, change, or feedback given by an individual or company to the project.

There are some ideas on how you can contribute:

- :ref:`Contributing to the codebase<Contributing to codebase>`.
- :ref:`Contributing to the code developer experience<Contributing to dev exp>`.
- :ref:`Contributing to the user experience<Contributing to user exp>`.
- :ref:`Contributing to the community<Contributing to community>`.
- :ref:`Other kinds of contributions<Other kinds of contributions>`.

> 

Author: Fixed, thanks!

Most types of contribution do not require programming skills.

.. _Contributing to codebase:

Contributing to the codebase
----------------------------

Contribution to software itself (its codebase) includes:

- `Proposing new features <https://docs.ansible.com/ansible/latest/community/reporting_bugs_and_features.html#requesting-a-feature>`_.
- `Reporting bugs <https://docs.ansible.com/ansible/latest/community/reporting_bugs_and_features.html#reporting-a-bug>`_.
- Participating in design and implementation of new components in dedicated issues, pull requests, GitHub discussions, and `IRC channels <https://docs.ansible.com/ansible/devel/community/communication.html#irc-channels>`_.
- Fixing bugs.
- Reviewing code.
- Refactoring code.
- Improving stability of an existing codebase through improving a `unit <https://docs.ansible.com/ansible/latest/dev_guide/developing_collections_testing.html#adding-unit-tests>`_ and `integration <https://docs.ansible.com/ansible/latest/dev_guide/developing_collections_testing.html#adding-integration-tests>`_ testing codebase and infrastructure.

> - Maintaining repositories (for example, backporting merged pull requests and releasing).

Reviewer: ```suggestion
- Maintaining repositories (for example, `backporting merged pull requests <https://docs.ansible.com/ansible/latest/community/development_process.html#backporting-merged-prs-in-ansible-core>`_ and releasing).
```

You can contribute to a codebase even if you do not have programming skills. For example, you can review proposals, test pull requests locally, add integration tests, file bug reports, create feature requests, or provide feedback to programmers in issues and pull requests.

See also:

- `Contributing to a codebase guide <contributing.rst>`_ to learn how to contribute to collection codebases.
- `Create a pull request quick-start guide <create_pr_quick_start_guide.rst>`_.
- `Review checklist <review_checklist.rst>_`.
- `Test a pull request locally guide <test_pr_locally_guide.rst>`_.
- `Collection maintainer guidelines <maintaining.rst>`_.
- `Collection releasing guidelines <releasing_collections.rst>`_.

> 

Reviewer: I don't know enough about github rst to know if these links will work, bu the 'Review Checklist' item looks different from the rest so you may want to review that one a bit.

Author: yeah, thanks! Fixed

- Writing a blog post about benefits and possible drawbacks of the project compared to other similar projects.
- Making a presentation on another IT community's event showing how the project can be used by that community to solve their issues.
- `Suggesting <https://github.com/ansible-community/community-topics/>`_ ideas on how the process of decision making can be improved.

In a broader sense, contribution to a community includes the activities mentioned in all the other types of contribution when they imply good interaction between participants or good impact on the community health and growth.

Examples:

- If you write a guide which significantly reduces a threshold of the first patch, it can lead to increase of a number of code contributors, which is definitely a contribution to the community.
- `Active participation in IRC / mailing lists / other discussions <https://docs.ansible.com/ansible/devel/community/communication.html>`_.
- Leading a blog or a website dedicated to the project.
- `Proposing and discussing ideas in the community meetings <https://github.com/ansible-community/community-topics/>`_, voting on proposals.

.. _Other kinds of contributions:

> Other kinds of contribution

Reviewer: ```suggestion
Other kinds of contributions
```

***********************
Contributing to Ansible
***********************

This document describes ways how you can contribute to the Open Source projects of the Ansible ecosystem.

> .. contents:: Topics

Reviewer 2: Our style guide suggests no heading for a local TOC - see https://docs.ansible.com/ansible/latest/dev_guide/style_guide/index.html#adding-local-tocs.
```suggestion
.. contents::
   :local:
```

What is a contribution to an Open Source project
================================================

> A contribution to an Open Source project is any good given by an individual or a company to the project including increasing its functionality, improvements of the codebase, code developer experience, user experience, increasing the project's popularity and community, providing infrastructure, and any other kinds of active social or material support.

Author: @acozine I think a contribution is broader than additions or changes (e.g. as described at the end - increasing popularity by writing something good about, providing infrastructure, etc.)

> A contribution to an Open Source project is any good given by an individual or a company to the project including increasing its functionality, improvements of the codebase, code developer experience, user experience, increasing the project's popularity and community, providing infrastructure, and any other kinds of active social or material support.

Author: so, imo, we should use something more abstract like `good` but maybe something else instead with similar meaning

> A contribution to an Open Source project is any good given by an individual or a company to the project including increasing its functionality, improvements of the codebase, code developer experience, user experience, increasing the project's popularity and community, providing infrastructure, and any other kinds of active social or material support.

Author: @acozine , changed to `any addition or positive change`

> Most the types of contribution to an Open Source project do not require programming skills.

Author: I'm speaking about the things mentioned above, so that's why i use `the`. But i think the suggested generalized form w/o `the` should also work. Thanks, committed

Classification of contributions
===============================

Contributions can be one of the following types:

>   - :ref:`Contribution to a codebase<Contribution to codebase>`.

Reviewer 2: This list repeats the items from the paragraph above. I'd remove the paragraph and the extra heading, and head this list "Some ideas for contributing to an open-source project:"

Author: Removed, thanks

  - :ref:`Contribution to code developer experience<Contribution to dev exp>`.
  - :ref:`Contribution to user experience<Contribution to user exp>`.
  - :ref:`Contribution to a community<Contribution to community>`.
  - :ref:`Other kinds of contributions<Other kinds of contributions>`.

.. _Contribution to codebase:

Contribution to a codebase
--------------------------

Contribution to software itself (its codebase) includes:

  - Proposing new features.
  - Reporting bugs.
  - Participating in design and implementation of new components.
  - Bug fixes.

>   - Code reviewing.

Reviewer 2: The other items are all phrased as "verb-ing noun" (for example, "reporting bugs"). For consistency, I'd do:
```suggestion
  - Reviewing code.
```

>   - Code refactoring.

Reviewer 2: Same here:
```suggestion
  - Refactoring code.
```

  - Improving stability of an existing codebase through improving a testing codebase and infrastructure.
  - Maintenance (for example, backporting and releasing).

> Examples of codebase contributions which do not require programming skills in scope of the Ansible project can be reviewing proposals, testing pull requests locally, adding integration tests, bug reports, feature requests, providing user experience to programmers in issues and pull requests.

Reviewer 2: ```suggestion
You can contribute to a codebase even if you do not have programming skills. For example, you can review proposals, test pull requests locally, add integration tests, file bug reports, create feature requests, or provide feedback to programmers in issues and pull requests.
```

Author: Much better:) thanks, committed

.. _Contribution to dev exp:

> Contribution to code developer experience

Reviewer 2: Thinking about this, I'd also make all the headers active verbs instead of nouns: `Contributing to the developer experience` and maybe `Finding the right contribution type for you` or something.

Author: Sounds good to me, changed, thanks!

-----------------------------------------

Contribution to code developer experience includes sharing experience through improving code development documentation, writing development guidelines, mentoring and training code contributors.

.. _Contribution to user exp:

Contribution to user experience
-------------------------------

Contribution to user experience includes sharing experience about how to use software more efficiently through improving user documentation, writing manuals and books, making videos, courses, and so on.

Examples:

  - Clarification and updating user documentation.
  - Making a presentation for a local meetup covering not well-known but useful usage aspects.

.. _Contribution to community:

Contribution to a community
-----------------------------

In a narrow sense, contribution to a community includes:

  - Promoting the project among people.
  - Thinking of and proposing ideas about how community processes can be improved.
  
Examples:

  - Organizing a meetup.

>   - Writing a blogpost about benefits of the project comparing with other similar projects.

Reviewer 3: There is a reason other similar projects exist, if a blog post is all talking good about a project I do not consider it helpful really.
```suggestion
  - Writing a blog post about benefits (and possible drawbacks) of the project comparing other similar projects.
```

Reviewer 3: I'm not native English speaker, but it seems that comparing to is not right (https://dictionary.cambridge.org/dictionary/english/comparing).

Author: sounds sensible, changed. though the balance is important if we want to see community growth, not a decline:D

Author: > I'm not a native English speaker, but it seems that "comparing to" is not right (https://dictionary.cambridge.org/dictionary/english/comparing).

Changed, thanks!

  - Making a presentation on another IT community's event showing how the project can be used by that community to solve their issues.
  - Suggesting ideas on how the process of decision making can be improved.

In a broader sense, contribution to a community includes the activities mentioned in all the other types of contribution when they imply good interaction between participants or good impact on the community health and growth.

Examples:

  - If you write a guide which significantly reduces a threshold of the first patch, it can lead to increase of a number of code contributors, which is definitely a contribution to the community.

>   - Active participation in IRC / mail / other discussions.

Reviewer 3: ```suggestion
  - Active participation in IRC / mailing lists / other discussions.
```

>   - Active participation in IRC / mail / other discussions.

Reviewer 3: Or
```suggestion
  - Active participation using IRC / mail / other methods.
```

Author: I'd like to highlight the discussions here. It implies even in-person discussions in meetups and any other kinds of discussions. So i'd leave `discussions` here:)
